### PR TITLE
feat: refactor training

### PR DESCRIPTION
**Pull request by stephantul**
This PR refactors a large part of the training logic. The base class is now much more feature complete, and takes care of everything that does not have labeled data. The classifier has become much smaller, and only takes care of data that has labels.

Here's a concise summary of the changes:
* We can now finetune static models on vectors. This is called `StaticModelForSimilarity`. This currently only supports cosine similarity training.
* The lightning modules for the classifier and multilabel classifier have been split up, and I added a new lightning module for similarity. The classifier lightning modules derive from the similarity ones
* Much of the logic has been consolidated in the base class.
* I added a `normalize` argument to the initializer. This controls whether the _model_ encoding is normalized.
* I changed the named arguments in `from_pretrained` to mimic those in `StaticModel` and added a deprecation warning to the old argument.
* I added a new argument called "balanced" to class_weight, added support for dicts, and deprecated support for tensors.
* I added a new argument called "validation_steps" to the fit method, which controls how many steps are between validation splits.
* `StaticModelPipeline` now supports projection, backed by a scikit learn `MLPRegressor`. This allows us to create static models with additional dense layers and put them in sklear pipelines.

And maybe some other changes I forgot. Not everything is fully covered by tests, but most of it is.

### CODE REVIEW

File: model2vec/train/lightning_modules.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        return loss
+
+
+class MultiLabelClassifierLightningModule(StaticLightningModule):
+    def __init__(self, model: nn.Module, learning_rate: float, class_weight: torch.Tensor | None = None) -> None:
+        """Initialize the LightningModule."""
+        super().__init__(model, learning_rate)
+        self.model = model
+        self.learning_rate = learning_rate
+        self.loss_function = nn.BCEWithLogitsLoss(weight=class_weight)
```
Comment (Pringled) [inline model2vec/train/lightning_modules.py]:
```suggestion
        self.loss_function = nn.BCEWithLogitsLoss(pos_weight=class_weight)
```
I think this should stay pos_weight like in the original code? Since it's per class

File: model2vec/train/base.py [hunk 2]
```diff
@@ -53,6 +74,8 @@ def __init__(
             self.vectors = vectors.float()
 
         if token_mapping is not None:
+            if len(token_mapping) != len(vectors):
```
Comment (Pringled) [inline model2vec/train/base.py]: I think this will break on vocab quantized models?

File: model2vec/train/base.py [hunk 3]
```diff
@@ -70,20 +93,56 @@ def construct_weights(self) -> nn.Parameter:
...
 
     @classmethod
     def from_pretrained(
-        cls: type[ModelType], *, out_dim: int = 2, model_name: str = "minishlab/potion-base-32m", **kwargs: Any
+        cls: type[ModelType], path: str = "minishlab/potion-base-32m", *, token: str | None = None, **kwargs: Any
     ) -> ModelType:
         """Load the model from a pretrained model2vec model."""
-        model = StaticModel.from_pretrained(model_name)
-        return cls.from_static_model(model=model, out_dim=out_dim, **kwargs)
+        if model_name := kwargs.pop("model_name"):
```
Comment (Pringled) [inline model2vec/train/base.py]:
```suggestion
        if model_name := kwargs.pop("model_name", None):
```
This gives a KeyError now if model_name is not passed, should just default to None

File: model2vec/inference/model.py [hunk 4]
```diff
@@ -23,25 +24,26 @@
...
-        return self.head.classes_
+
+        last_head = self.head[-1]
+        self.classes_: None | np.ndarray = None
+        if isinstance(last_head, MLPClassifier):
+            activation = last_head.out_activation_
+            self.classifier_type = HeadType.MULTILABEL if activation == "logistic" else HeadType.CLASSIFIER
+            self.classes_ = self.head.classes_
+        else:
+            self.classifier_type = HeadType.PROJECTOR
```
Comment (Pringled) [inline model2vec/inference/model.py]: Hmm I'm not 100% sure about this; a classifier can technically be anything right? So if we use a logreg instead of MLP, it will incorrectly become a projector type since it only checks for MLPClassifier? But mb I'm overthinking this since we don't support other classification heads anyway so this would only happen if someone actually modifies the code

File: model2vec/train/base.py [hunk 5]
```diff
@@ -1,48 +1,69 @@
...
-from model2vec.train.utils import get_probable_pad_token_id
+from model2vec.inference import StaticModelPipeline
+from model2vec.train.dataset import TextDataset
+from model2vec.train.utils import get_probable_pad_token_id, to_pipeline, train_test_split
 
 logger = logging.getLogger(__name__)
 
 
-class FinetunableStaticModel(nn.Module):
+class _BaseFinetuneable(nn.Module):
```
Comment (Pringled) [inline model2vec/train/base.py]: Should this be made private? It replaces FinetunableStaticModel which was public; I think public is more correct as someone can inherit ffrom this if they want to add custom tasks

File: model2vec/train/classifier.py [hunk 6]
```diff
@@ -3,32 +3,28 @@
...
 
 logger = logging.getLogger(__name__)
+
 _DEFAULT_RANDOM_SEED = 42
+LabelType = list[str] | list[list[str]]
 
-LabelType = TypeVar("LabelType", list[str], list[list[str]])
 
+class StaticModelForClassification(_BaseFinetuneable):
+    val_metric = "val_accuracy"
```
Comment (Pringled) [inline model2vec/train/classifier.py:25]: Should we also set these explicitly for similarity models? I do like having them as class variables

**Review by Pringled — CHANGES_REQUESTED** [review]
Few small 🐛 but mostly good to go, very nice

**Review by Pringled — APPROVED** [review]
LGTM!
